crates/interpreter/src/syntax.rs

```diff
@@ -546,6 +557,47 @@ impl CvtOp {
     }
 }
 
+#[cfg(feature = "threads")]
+macro_rules! impl_atomic_op {
+    ($u:ident) => {
+        paste::paste! {
+            impl AtomicUnOp {
+                pub fn $u(&self, mem: *mut $u, c: $u) -> $u {
+                    let atomic_val = unsafe { [< Atomic $u:upper >]::from_ptr(mem) };
+                    match self {
+                        AtomicUnOp::Add => atomic_val.fetch_add(c, Ordering::SeqCst),
+                        AtomicUnOp::Sub => atomic_val.fetch_sub(c, Ordering::SeqCst),
+                        AtomicUnOp::And => atomic_val.fetch_and(c, Ordering::SeqCst),
+                        AtomicUnOp::Or => atomic_val.fetch_or(c, Ordering::SeqCst),
+                        AtomicUnOp::Xor => atomic_val.fetch_xor(c, Ordering::SeqCst),
+                        AtomicUnOp::Xchg => atomic_val.swap(c, Ordering::SeqCst),
+                    }
+                }
+            }
+            impl AtomicBinOp {
+                pub fn $u(&self, mem: *mut $u, c: $u, c2: $u) -> $u {
+                    let atomic_val = unsafe { [< Atomic $u:upper >]::from_ptr(mem) };
+                    match atomic_val.compare_exchange(c2, c, Ordering::SeqCst, Ordering::SeqCst) {
+                                Err(v) => v,
+                                Ok(v) => v
+                    }
+                }
+            }
+        }
+    };
+}
+
+#[cfg(feature = "threads")]
+impl_atomic_op!(i64);
+#[cfg(feature = "threads")]
+impl_atomic_op!(i32);
```

Comment:
I don't think the spec uses signed integers for atomics, because all operations are sign-agnostics. I guess the confusion may come from the nomenclature between Rust and Wasm:
- `iNN` means signed integer in Rust but it means "no-signedness" integer in Wasm, see https://webassembly.github.io/spec/core/syntax/values.html#integers
- `uNN` means unsigned integer in both Rust and Wasm
- `sNN` means signed integer in Wasm

Reply:
Yeah, thanks for the pointers. I was indeed thinking `iNN` is signed in Rust and wasm. I'm now using unsigned atomics everywhere.

IIUC this works because `iNN` in wasm can be signed or unsigned, but is in twos complement in the signed case so unsigned addition will work just fine. And we always zero-extend by definition.